### PR TITLE
DR2-2318 Changes for DRI migration

### PR DESCRIPTION
**Pull request by MancunianSam**
* Add a filePath parameter. This comes from TDR and from the DRI migration script. This goes into the asset opex
* Allow multiple checksums, either as SHA256ServerSideChecksum from TDR or from a set of checksum_<algorithm> keys from DRI
* Rename the package builder classes and variables to remove tdr. I'll change the package names later.
* Change the package builder so it isn't streaming the metadata files. It's not necessary as they're never that large and it's confusing.
* Add an optional `driBatchReference` parameter which comes from the DRI migration. If it's populated, it's added as an identifier.
* Pass in source system as an environment variable to the package builder
* Fiddled the build.sbt so that it will create two jars with different names for the package builder and aggregator.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -31,6 +31,7 @@ jobs:
           (cd ./python/lambdas/copy-files-from-tdr/ && bash deploy.sh)
           zip -jA copy-files-from-tdr python/lambdas/copy-files-from-tdr/lambda_function.py
           cp copy-files-from-tdr target/outputs
+          cp copy-files-from-tdr target/outputs/copy-files-from-dri
```
Comment (techncl) [inline .github/workflows/build.yml:34]: How come you are copying this to two places?

File: scala/lambdas/utils/src/main/scala/uk/gov/nationalarchives/utils/ExternalUtils.scala [hunk 2]
```diff
@@ -203,12 +207,28 @@ object ExternalUtils {
...
+    if keys.contains("SHA256ServerSideChecksum") then
+      c.downField("SHA256ServerSideChecksum").as[String].map { sha256Checksum =>
+        List(Checksum("sha256", sha256Checksum))
+      }
+    else if checksumKeys.nonEmpty then
+      checksumKeys.traverse { key =>
+        val algorithm = key.replace(checksumPrefix, "")
+        c.downField(key).as[String].map(fingerprint => Checksum(algorithm, fingerprint))
+      }
+    else
```
Comment (techncl) [inline scala/lambdas/utils/src/main/scala/uk/gov/nationalarchives/utils/ExternalUtils.scala:229]: Could you not just do

```
    if keys.contains("SHA256ServerSideChecksum") then
      c.downField("SHA256ServerSideChecksum").as[String].map { sha256Checksum =>
        List(Checksum("sha256", sha256Checksum))
      }
    else
      checksumKeys.traverse { key =>
        val algorithm = key.replace(checksumPrefix, "")
        c.downField(key).as[String].map(fingerprint => Checksum(algorithm, fingerprint))
      }

```?

**Review by techncl — COMMENTED** [review]
Looks good! Just some minor comments. Also should we add tests for `create_ingest_metadata.py`?

File: python/dri-migration/create_ingest_metadata.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    path = f"{os.environ['DROID_PATH']}\\{prefix}"
+
+    pattern = re.compile(r'((x-)?fmt-\d{1,5})-signature-id-\d{1,5}(\.[A-Za-z]{1,10})')
+
+    directory_list = listdir(path)
+
+    for name in directory_list:
+        match = pattern.search(name)
+        if match:
+            puid = match.group(1).replace('fmt-', 'fmt/')
```
Comment (techncl) [inline python/dri-migration/create_ingest_metadata.py:28]: Do you think it's worth doing this `.replace(f'{prefix}-', f'{prefix}/')` instead?

File: python/dri-migration/create_ingest_metadata.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+create_skeleton_suite_lookup('x-fmt')
+
+with open("ingest_query.sql") as query:
+    sql = query.read()
+    cur.execute(sql)
+
+
+account_number = os.environ["ACCOUNT_NUMBER"]
+environment = os.environ["ENVIRONMENT"]
+bucket = f"{environment}-dr2-ingest-raw-cache"
```
Comment (techncl) [inline python/dri-migration/create_ingest_metadata.py:42]: Should this be an environment variable?

File: scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/TestUtils.scala [hunk 5]
```diff
@@ -78,12 +95,14 @@ object TestUtils:
 
       override def upload(bucket: String, key: String, publisher: Publisher[ByteBuffer]): IO[CompletedUpload] = {
         if uploadError then IO.raiseError(new Exception(s"Error uploading $key to $bucket"))
-        else
+        else {
           for {
             jsonString <- publisher.toStreamBuffered[IO](1024).map(_.array().map(_.toChar).mkString).compile.string
             json <- IO.fromEither(decode[List[MetadataObject]](jsonString))
+            _ <- IO.println(jsonString)
```
Comment (techncl) [inline scala/lambdas/preingest-tdr-package-builder/src/test/scala/uk/gov/nationalarchives/preingesttdrpackagebuilder/TestUtils.scala:102]: Is this supposed to be here?

File: python/dri-migration/create_ingest_metadata.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    if not rows:
+        break
+    for row in rows:
+        puid = row[key_indexes["PUID"]]
+        asset_uuid = row[key_indexes["UUID"]]
+        file_id = row[key_indexes["FILEID"]]
+        file_path = row[key_indexes["FILE_PATH"]]
+        full_path = row[key_indexes["FULLPATH"]]
+        checksum_data = json.loads(row[key_indexes["FIXITIES"]])
+        metadata = {
```
Comment (techncl) [inline python/dri-migration/create_ingest_metadata.py:85]: Should we use the metadata schema to validate that we have all of the necessary fields?

**Comment by techncl** [comment]
Could you add a README for the DRI migration project?

File: build.sbt [hunk 7]
```diff
@@ -1,11 +1,12 @@
 import Dependencies.*
-ThisBuild / scalaVersion := "3.7.1"
+ThisBuild / scalaVersion := "3.7.0"
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply build.sbt:2]: I was fiddling with the code analysis tools and it didn't have a version for 3.7.1 I'm not doing that now so I've put it back.

File: python/dri-migration/create_ingest_metadata.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+conn = oracledb.connect(dsn='localhost/SDB4', user="STORE", password=os.environ['STORE_PASSWORD'])
+cur = conn.cursor()
+
+puid_lookup = {}
+
+
+def create_skeleton_suite_lookup(prefix):
+    path = f"{os.environ['DROID_PATH']}\\{prefix}"
+
+    pattern = re.compile(r'((x-)?fmt-\d{1,5})-signature-id-\d{1,5}(\.[A-Za-z]{1,10})')
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply python/dri-migration/create_ingest_metadata.py:21]: 1. This isn't a long regex. I did try it but I don't think it helped readability that much. 
2. I tried this and it doesn't seem to work, it doesn't find any matches like this.

File: python/dri-migration/create_ingest_metadata.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+account_number = os.environ["ACCOUNT_NUMBER"]
+environment = os.environ["ENVIRONMENT"]
+bucket = f"{environment}-dr2-ingest-raw-cache"
+queue_url = f"https://sqs.eu-west-2.amazonaws.com/{account_number}/{environment}-dr2-copy-files-from-dri"
+
+page_size = 100
+
+s3_client = boto3.client("s3")
+sqs_client = boto3.client("s3")
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply python/dri-migration/create_ingest_metadata.py]: Yeah it should, I've changed it.

File: scala/lambdas/build.sbt [hunk 10]
```diff
@@ -267,33 +284,56 @@ lazy val postIngestStateChangeHandler = (project in file("postingest-state-chang
...
+  scalaCheckPlus % Test
+)
+
 lazy val preIngestTdrPackageBuilder = (project in file("preingest-tdr-package-builder"))
+  .settings(name := baseDirectory.value.getName)
   .settings(commonSettings)
   .dependsOn(utils, dynamoFormatters)
+  .settings(packageBuilderSettings)
+
+lazy val preingestDriPackageBuilder = (project in file("preingest-tdr-package-builder"))
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/build.sbt:304]: Yeah, using the same source code but for a different module. I'll rename the directory to `preingest-package-builder` at some point but not now.

File: scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json [hunk 11]
```diff
@@ -18,8 +18,7 @@
       }
     },
     "transferringBody": {
-      "type": "string",
-      "minLength": 1
+      "type": "string"
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json:21]: No, that's not right. I think I thought the min length was making it mandatory but that's only if populated. I've put the minLength back.

[the remark this answers is not in the document]
Comment (techncl) [reply scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json:21]: Ah, I see. Could we add a test that checks this?

[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json]: I've just tried this with the minLength and no transferringBody key and it fails which we don't want it to so I've removed the minLength again.

[the remark this answers is not in the document]
Comment (techncl) [reply scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json:21]: That's weird. Wait, is this happening because you've added `packageMetadata.transferringBody.getOrElse("")`? i.e. even if the transferring body is missing we'll still make it an empty string?

**Comment by MancunianSam** [comment]
I've added a README. 

I've also made the source system an enum and changed the case class fields for `PackageMetadata` so the case of the fields is consistent now. It's all in this commit https://github.com/nationalarchives/dr2-ingest/pull/586/commits/12c650d06f3a76c1dc2f4cbc9d56958b8726036d

File: scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json:21]: Oh yeah, that'll be it. I was being lazy, I'll make the asset field optional.

[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/resources/asset/asset-validation-schema.json]: I've modified one of the validator tests to remove the transferring body field and added another test to check we get an error if it's there but empty and this seems to work.

File: scala/lambdas/preingest-tdr-package-builder/README.md [hunk 12]
```diff
@@ -52,3 +54,4 @@ The lambda outputs this json.
 | LOCK_DDB_TABLE_GROUPID_GSI_NAME | The lock table global secondary index name in Dynamo                                                                                    |
 | OUTPUT_BUCKET_NAME              | The bucket the TDR file and metadata file are stored in                                                                                 |
 | CONCURRENCY                     | An optional parameter which configures how many items the lambda processes in parallel. Used for performance tweaking. Defaults to 1000 |
+| SOURCE_SYSTEM                   | TDR or DRI                                                                                                                              |
```
Comment (techncl) [inline scala/lambdas/preingest-tdr-package-builder/README.md:57]: This is missing the TRE source system

**Comment by MancunianSam** [comment]
I've added a definitely not copilot generated python test for the migration script. It only does the happy path but that's fine, I don't want to get bogged down checking for errors.

File: python/dri-migration/create_ingest_metadata_test.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            mock_create_skeleton, _, __, mock_connect,
+    ):
+        os.environ['CLIENT_LOCATION'] = '/test/client'
+        os.environ['STORE_PASSWORD'] = 'password'
+        os.environ['DROID_PATH'] = '/test/droid'
+        os.environ['ACCOUNT_NUMBER'] = '123456789'
+        os.environ['ENVIRONMENT'] = 'testenv'
+
+        mock_create_skeleton.return_value = {
+            "fmt/123": {"file_path": "/test/file1"}
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply python/dri-migration/create_ingest_metadata_test.py]: Yeah ok. I think the easiest way is to have to rows, one which returns fmt/123 and one for x-fmt/123. I've changed the ids slightly on the rows to test that we're getting the right data from multiple rows. 

I've also added in a test for `create_skeleton_suite_lookup` and `calculate_checksum` and these do check the slightly unhappy path.
